cdap-metrics-sidecar/src/test/java/io/cdap/cdap/jmx/metrics/JMXMetricsCollectorTest.java

```diff
@@ -0,0 +1,136 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.jmx.metrics;
+
+import com.google.common.collect.ImmutableMap;
+import io.cdap.cdap.api.Environment;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.api.metrics.MetricsContext;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.net.MalformedURLException;
+import java.rmi.registry.LocateRegistry;
+import java.util.Map;
+import javax.management.MBeanServer;
+import javax.management.remote.JMXConnectorServer;
+import javax.management.remote.JMXConnectorServerFactory;
+import javax.management.remote.JMXServiceURL;
+
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class JMXMetricsCollectorTest {
+  private static final int SERVER_PORT = 11023;
+  private static final String SERVICE_NAME = "test-service";
+  private static JMXConnectorServer svr;
+
+  private static final Map<String, String> METRICS_CONTEXT = ImmutableMap.of(
+    Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+    Constants.Metrics.Tag.COMPONENT, SERVICE_NAME);
+
+  @Mock
+  private MetricsCollectionService mockMetricsService;
+  @Mock
+  private MetricsContext mockContext;
+  @Mock
+  private Environment mockEnv;
+
+  @Before
+  public void beforeEach() {
+    MockitoAnnotations.initMocks(this);
+    when(mockMetricsService.getContext(METRICS_CONTEXT)).thenReturn(mockContext);
+    when(mockEnv.getVariable("HOSTNAME")).thenReturn("test-host");
+  }
+
+  @BeforeClass
+  public static void setupClass() throws IOException {
+    svr = createJMXConnectorServer(SERVER_PORT);
+    svr.start();
+    Assert.assertEquals(svr.isActive(), true);
+  }
+
+  @AfterClass
+  public static void teardownClass() throws IOException {
+    svr.stop();
+  }
+
+  @Test
+  public void testInvalidPortInConfig() throws InterruptedException, MalformedURLException {
+    when(mockEnv.getVariable("SERVICE_NAME")).thenReturn(SERVICE_NAME);
+    CConfiguration cConf = CConfiguration.create();
+    cConf.setInt(Constants.JMXMetricsCollector.POLL_INTERVAL, 100);
+    cConf.setInt(Constants.JMXMetricsCollector.SERVER_PORT, -1);
+    JMXMetricsCollector jmxMetrics = new JMXMetricsCollector(cConf, mockMetricsService, mockEnv);
+    jmxMetrics.start();
+    Thread.sleep(200);
```

Comment:
Avoid using Thread.sleep() directly in tests, better to wait on a specific value relevant for the test case. See https://github.com/cdapio/cdap/blob/b2ea6755f7b384a24d4d3cff9b82e445abf74905/cdap-common/src/main/java/io/cdap/cdap/common/utils/Tasks.java#L49

Reply:
I want to test that the number of metrics emitted per unit time is as expected.  If I waitFor 4 metrics to be emitted,  then I'll still need to check that it did not emit those metrics too fast, say in 30 millisecs instead of 300 millisecs. So I want to wait for exactly 350 millisecs. 
Is there a way to achieve this using waitFor or can I model the test in some other way?

Reply:
Thread.sleep calls removed.